formats/index/decoder.go

```diff
@@ -0,0 +1,389 @@
+	}
+
+	var base string
+	if d.lastEntry != nil {
+		base = d.lastEntry.Name[:len(d.lastEntry.Name)-int(l)]
+	}
+
+	name, err := readUntil(d.r, 0)
+	if err != nil {
+		return "", err
+	}
+
+	return base + string(name), nil
+}
+
+func (d *Decoder) doReadEntryName(e *Entry) (string, error) {
+	pLen := e.Flags & nameMask
+
+	name := make([]byte, int64(pLen))
+	if err := binary.Read(d.r, binary.BigEndian, &name); err != nil {
+		return "", err
+	}
+
+	return string(name), nil
+}
+
+// Index entries are padded out to the next 8 byte alignment
+// for historical reasons related to how C Git read the files.
+func (d *Decoder) padEntry(idx *Index, e *Entry, read int) error {
+	if idx.Version == 4 {
+		return nil
+	}
+
+	entrySize := read + len(e.Name)
+	padLen := 8 - entrySize%8
+	if _, err := io.CopyN(ioutil.Discard, d.r, int64(padLen)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *Decoder) readExtensions(idx *Index) error {
+	var err error
+	for {
+		err = d.readExtension(idx)
+		if err != nil {
+			break
+		}
+	}
+
+	if err == io.EOF {
+		return nil
+	}
+
+	return nil
+}
+
+func (d *Decoder) readExtension(idx *Index) error {
+	var s = make([]byte, 4)
+	if _, err := d.r.Read(s); err != nil {
+		return err
+	}
+
+	var len uint32
+	if err := binary.Read(d.r, binary.BigEndian, &len); err != nil {
+		return err
+	}
+
+	switch {
+	case bytes.Equal(s, treeExtSignature):
+		t := &Tree{}
+		td := &TreeExtensionDecoder{&io.LimitedReader{R: d.r, N: int64(len)}}
+		if err := td.Decode(t); err != nil {
+			return err
+		}
+
+		idx.Cache = t
+	}
+
+	return nil
+}
+
+func validateHeader(r io.Reader) (version uint32, err error) {
+	var s = make([]byte, 4)
+	if _, err := r.Read(s); err != nil {
+		return 0, err
+	}
+
+	if !bytes.Equal(s, indexSignature) {
+		return 0, ErrMalformedSignature
+	}
+
+	if err := binary.Read(r, binary.BigEndian, &version); err != nil {
+		return 0, err
+	}
+
+	if version < IndexVersionSupported.Min || version > IndexVersionSupported.Max {
+		return 0, ErrUnsupportedVersion
+	}
+
+	return
+}
+
+type TreeExtensionDecoder struct {
+	r io.Reader
+}
+
+func (d *TreeExtensionDecoder) Decode(t *Tree) error {
+	for {
+		e, err := d.readEntry()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+
+			return err
+		}
+
+		t.Entries = append(t.Entries, *e)
+	}
+}
+
+func (d *TreeExtensionDecoder) readEntry() (*TreeEntry, error) {
+	e := &TreeEntry{}
+
+	path, err := readUntil(d.r, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	e.Path = string(path)
+
+	count, err := readUntil(d.r, ' ')
+	if err != nil {
+		return nil, err
+	}
+
+	i, err := strconv.Atoi(string(count))
+	if err != nil {
+		return nil, err
+	}
+
+	e.Entries = i
+
+	trees, err := readUntil(d.r, 10)
+	if err != nil {
+		return nil, err
+	}
+
+	i, err = strconv.Atoi(string(trees))
+	if err != nil {
+		return nil, err
+	}
+
+	e.Trees = i
+
+	if err := binary.Read(d.r, binary.BigEndian, &e.Hash); err != nil {
+		return nil, err
+	}
+
+	return e, nil
```

Comment:
Can you replace all this function with `fmt.Fscanf(d.r, "%s\x00%d %s\n%s" &e.Path, &e.Entries, &e.Tree, &e.Hash)` ? or some other one-liner?